zhaquirks/xiaomi/aqara/smoke.py

```diff
@@ -0,0 +1,131 @@
+"""Quirk for LUMI lumi.sensor_smoke.acn03 smoke sensor."""
+from typing import Any
+
+from zigpy.profiles import zha
+from zigpy.quirks import CustomDevice
+import zigpy.types as types
+from zigpy.zcl.clusters.general import Basic, Identify, Ota, PowerConfiguration
+from zigpy.zcl.clusters.security import IasZone
+from zigpy.zdo.types import NodeDescriptor
+
+from zhaquirks import Bus, LocalDataCluster
+from zhaquirks.const import (
+    DEVICE_TYPE,
+    ENDPOINTS,
+    INPUT_CLUSTERS,
+    MODELS_INFO,
+    NODE_DESCRIPTOR,
+    OUTPUT_CLUSTERS,
+    PROFILE_ID,
+    ZONE_STATUS,
+)
+from zhaquirks.xiaomi import LUMI, XiaomiAqaraE1Cluster, XiaomiPowerConfiguration
+
+BUZZER_MANUAL_MUTE = 0x0126
+SELF_TEST = 0x0127
+SMOKE = 0x013A
+SMOKE_DENSITY = 0x013B
+HEARTBEAT_INDICATOR = 0x013C
+BUZZER_MANUAL_ALARM = 0x013D
+BUZZER = 0x013E
+LINKAGE_ALARM = 0x014B
+LINKAGE_ALARM_STATE = 0x014C
+SMOKE_DENSITY_DBM = 0x1403  # fake attribute for smoke density in dB/m
+
+SMOKE_DENSITY_DBM_MAP = {
+    0: 0,
+    1: 0.085,
+    2: 0.088,
+    3: 0.093,
+    4: 0.095,
+    5: 0.100,
+    6: 0.105,
+    7: 0.110,
+    8: 0.115,
+    9: 0.120,
+    10: 0.125,
+}
+
+
+class OppleCluster(XiaomiAqaraE1Cluster):
+    """Opple cluster."""
+
+    ep_attribute = "opple_cluster"
+    attributes = {
+        BUZZER_MANUAL_MUTE: ("buzzer_manual_mute", types.uint8_t, True),
+        SELF_TEST: ("self_test", types.Bool, True),
+        SMOKE: ("smoke", types.uint8_t, True),
+        SMOKE_DENSITY: ("smoke_density", types.uint8_t, True),
+        HEARTBEAT_INDICATOR: ("heartbeat_indicator", types.uint8_t, True),
+        BUZZER_MANUAL_ALARM: ("buzzer_manual_alarm", types.uint8_t, True),
+        BUZZER: ("buzzer", types.uint32_t, True),
+        LINKAGE_ALARM: ("linkage_alarm", types.uint8_t, True),
+        LINKAGE_ALARM_STATE: ("linkage_alarm_state", types.uint8_t, True),
+        SMOKE_DENSITY_DBM: ("smoke_density_dbm", types.Single, True),
+    }
+
+    def _update_attribute(self, attrid: int, value: Any) -> None:
+        """Pass attribute update to another cluster if necessary."""
+        super()._update_attribute(attrid, value)
+        if attrid == SMOKE:
+            self.endpoint.ias_zone.update_attribute(ZONE_STATUS, value)
+        elif attrid == SMOKE_DENSITY:
+            self.update_attribute(SMOKE_DENSITY_DBM, SMOKE_DENSITY_DBM_MAP[value])
```

Comment:
Do we prefer an error here if `value` not in `SMOKE_DENSITY_DBM_MAP` or silently return something else?
Not an strong opinion from me, just thinking loud.

Reply:
Hmm. The sensor seems to only report a value from 0 to 10 for the smoke density (in my testing and Z2M does this in a very similar way: https://github.com/Koenkk/zigbee-herdsman-converters/blob/ac73a48e057ab5a0b7d8d55bbd09a6dda258be15/lib/xiaomi.js#L389-L390)

I think if it reports another value, we should get an exception, so some user hopefully reports this 😄